x-pack/plugins/index_management/public/application/sections/home/index_list/details_page/details_page_mappings_content.tsx

```diff
@@ -16,104 +16,180 @@ import {
+import { useMappingsStateListener } from '../../../../../application/components/mappings_editor/use_state_listener';
 import { useAppContext } from '../../../../app_context';
+import { documentationService } from '../../../../services';
 
-export const DetailsPageMappingsContent: FunctionComponent<{ index: Index; data: string }> = ({
-  index,
-  data,
-}) => {
+export const DetailsPageMappingsContent: FunctionComponent<{
+  index: Index;
+  data: string;
+  jsonData: any;
+}> = ({ index, data, jsonData }) => {
   const {
     services: { extensionsService },
     core: { getUrlForApp },
   } = useAppContext();
-  return (
-    // using "rowReverse" to keep docs links on the top of the mappings code block on smaller screen
-    <EuiFlexGroup
-      wrap
-      direction="rowReverse"
+  const { euiTheme } = useEuiTheme();
+
+  const [toggleOn, setToggleOn] = useState(true);
+  const onToggleChange = () => {
+    setToggleOn((prevValue) => !prevValue);
+  };
+  const { parsedDefaultValue } = useMemo<MappingsEditorParsedMetadata>(() => {
+    const mappingsDefinition = extractMappingsDefinition(jsonData);
+
+    if (mappingsDefinition === null) {
+      return { multipleMappingsDeclared: true };
+    }
+
+    const {
+      _source,
+      _meta,
+      _routing,
+      _size,
+      dynamic,
+      properties,
+      runtime,
+      /* eslint-disable @typescript-eslint/naming-convention */
+      numeric_detection,
+      date_detection,
+      dynamic_date_formats,
+      dynamic_templates,
+      /* eslint-enable @typescript-eslint/naming-convention */
+    } = mappingsDefinition;
+
+    const parsed = {
+      configuration: {
+        _source,
+        _meta,
+        _routing,
+        _size,
+        dynamic,
+        numeric_detection,
+        date_detection,
+        dynamic_date_formats,
+      },
+      fields: properties,
+      templates: {
+        dynamic_templates,
+      },
+      runtime,
+    };
+
+    return { parsedDefaultValue: parsed, multipleMappingsDeclared: false };
+  }, [jsonData]);
+
+  useMappingsStateListener({ value: parsedDefaultValue, status: 'disabled' });
+
+  const {
+    fields: { byId, rootLevelFields },
+  } = useMappingsState();
+
+  const getField = useCallback((fieldId: string) => byId[fieldId], [byId]);
+  const fields = useMemo(() => rootLevelFields.map(getField), [rootLevelFields, getField]);
+
+  const jsonBlock = (
+    <EuiCodeBlock
+      language="json"
+      isCopyable
+      data-test-subj="indexDetailsMappingsCodeBlock"
       css={css`
         height: 100%;
       `}
     >
-      <EuiFlexItem
-        grow={1}
+      {data}
+    </EuiCodeBlock>
+  );
+
+  const treeViewBlock = <FieldsList fields={fields} />;
+
+  return (
+    // using "rowReverse" to keep docs links on the top of the mappings code block on smaller screen
+    <>
+      <EuiFlexGroup style={{ marginBottom: euiTheme.size.l }}>
+        <EuiButton onClick={onToggleChange}>{!toggleOn ? 'JSON' : 'Table View'}</EuiButton>
```

Comment:
Can we add translations/`FormattedMessage` for the "JSON" and "Table View" strings?

Reply:
@ElenaStoeva Can you please send me the mapping of index for which the page is blank? I want to regenerate the issue locally. Thank you.

Reply:
Hi @saikatsarkar056, it seems the issue is fixed now!

<details>
<summary>Here is what the mapping was:</summary>

```
{
    "mappings": {
        "dynamic": "false",
        "_data_stream_timestamp": {
            "enabled": true
        },
        "properties": {
            "@timestamp": {
                "type": "date"
            },
            "ecs": {
                "properties": {
                    "version": {
                        "type": "keyword",
                        "ignore_above": 1024
                    }
                }
            },
            "error": {
                "properties": {
                    "code": {
                        "type": "keyword",
                        "ignore_above": 1024
                    },
                    "id": {
                        "type": "keyword",
                        "ignore_above": 1024
                    },
                    "message": {
                        "type": "text",
                        "norms": false
                    },
                    "stack_trace": {
                        "type": "keyword",
                        "index": false,
                        "doc_values": false,
                        "ignore_above": 1024,
                        "fields": {
                            "text": {
                                "type": "text",
                                "norms": false
                            }
                        }
                    },
                    "type": {
                        "type": "keyword",
                        "ignore_above": 1024
                    }
                }
            },
            "event": {
                "properties": {
                    "action": {
                        "type": "keyword",
                        "ignore_above": 1024
                    },
                    "category": {
                        "type": "keyword",
                        "ignore_above": 1024,
                        "meta": {
                            "isArray": "true"
                        }
                    },
                    "code": {
                        "type": "keyword",
                        "ignore_above": 1024
                    },
                    "created": {
                        "type": "date"
                    },
                    "dataset": {
                        "type": "keyword",
                        "ignore_above": 1024
                    },
                    "duration": {
                        "type": "long"
                    },
                    "end": {
                        "type": "date"
                    },
                    "hash": {
                        "type": "keyword",
                        "ignore_above": 1024
                    },
                    "id": {
                        "type": "keyword",
                        "ignore_above": 1024
                    },
                    "ingested": {
                        "type": "date"
                    },
                    "kind": {
                        "type": "keyword",
                        "ignore_above": 1024
                    },
                    "module": {
                        "type": "keyword",
                        "ignore_above": 1024
                    },
                    "original": {
                        "type": "keyword",
                        "index": false,
                        "doc_values": false,
                        "ignore_above": 1024
                    },
                    "outcome": {
                        "type": "keyword",
                        "ignore_above": 1024
                    },
                    "provider": {
                        "type": "keyword",
                        "ignore_above": 1024
                    },
                    "reason": {
                        "type": "keyword",
                        "ignore_above": 1024
                    },
                    "reference": {
                        "type": "keyword",
                        "ignore_above": 1024
                    },
                    "risk_score": {
                        "type": "float"
                    },
                    "risk_score_norm": {
                        "type": "float"
                    },
                    "sequence": {
                        "type": "long"
                    },
                    "severity": {
                        "type": "long"
                    },
                    "start": {
                        "type": "date"
                    },
                    "timezone": {
                        "type": "keyword",
                        "ignore_above": 1024
                    },
                    "type": {
                        "type": "keyword",
                        "ignore_above": 1024,
                        "meta": {
                            "isArray": "true"
                        }
                    },
                    "url": {
                        "type": "keyword",
                        "ignore_above": 1024
                    }
                }
            },
            "kibana": {
                "properties": {
                    "action": {
                        "properties": {
                            "execution": {
                                "properties": {
                                    "gen_ai": {
                                        "properties": {
                                            "usage": {
                                                "properties": {
                                                    "completion_tokens": {
                                                        "type": "long"
                                                    },
                                                    "prompt_tokens": {
                                                        "type": "long"
                                                    },
                                                    "total_tokens": {
                                                        "type": "long"
                                                    }
                                                }
                                            }
                                        }
                                    },
                                    "source": {
                                        "type": "keyword",
                                        "ignore_above": 1024
                                    },
                                    "uuid": {
                                        "type": "keyword",
                                        "ignore_above": 1024
                                    }
                                }
                            },
                            "id": {
                                "type": "keyword",
                                "ignore_above": 1024
                            },
                            "name": {
                                "type": "keyword",
                                "ignore_above": 1024
                            }
                        }
                    },
                    "alert": {
                        "properties": {
                            "flapping": {
                                "type": "boolean"
                            },
                            "maintenance_window_ids": {
                                "type": "keyword",
                                "ignore_above": 1024,
                                "meta": {
                                    "isArray": "true"
                                }
                            },
                            "rule": {
                                "properties": {
                                    "consumer": {
                                        "type": "keyword",
                                        "ignore_above": 1024
                                    },
                                    "execution": {
                                        "properties": {
                                            "metrics": {
                                                "properties": {
                                                    "alert_counts": {
                                                        "properties": {
                                                            "active": {
                                                                "type": "long"
                                                            },
                                                            "new": {
                                                                "type": "long"
                                                            },
                                                            "recovered": {
                                                                "type": "long"
                                                            }
                                                        }
                                                    },
                                                    "claim_to_start_duration_ms": {
                                                        "type": "long"
                                                    },
                                                    "es_search_duration_ms": {
                                                        "type": "long"
                                                    },
                                                    "execution_gap_duration_s": {
                                                        "type": "long"
                                                    },
                                                    "number_of_delayed_alerts": {
                                                        "type": "long"
                                                    },
                                                    "number_of_generated_actions": {
                                                        "type": "long"
                                                    },
                                                    "number_of_searches": {
                                                        "type": "long"
                                                    },
                                                    "number_of_triggered_actions": {
                                                        "type": "long"
                                                    },
                                                    "persist_alerts_duration_ms": {
                                                        "type": "long"
                                                    },
                                                    "prepare_rule_duration_ms": {
                                                        "type": "long"
                                                    },
                                                    "process_alerts_duration_ms": {
                                                        "type": "long"
                                                    },
                                                    "process_rule_duration_ms": {
                                                        "type": "long"
                                                    },
                                                    "rule_type_run_duration_ms": {
                                                        "type": "long"
                                                    },
                                                    "total_enrichment_duration_ms": {
                                                        "type": "long"
                                                    },
                                                    "total_indexing_duration_ms": {
                                                        "type": "long"
                                                    },
                                                    "total_run_duration_ms": {
                                                        "type": "long"
                                                    },
                                                    "total_search_duration_ms": {
                                                        "type": "long"
                                                    },
                                                    "trigger_actions_duration_ms": {
                                                        "type": "long"
                                                    }
                                                }
                                            },
                                            "status": {
                                                "type": "keyword",
                                                "ignore_above": 1024
                                            },
                                            "status_order": {
                                                "type": "long"
                                            },
                                            "uuid": {
                                                "type": "keyword",
                                                "ignore_above": 1024
                                            }
                                        }
                                    },
                                    "revision": {
                                        "type": "long"
                                    },
                                    "rule_type_id": {
                                        "type": "keyword",
                                        "ignore_above": 1024
                                    }
                                }
                            },
                            "uuid": {
                                "type": "keyword",
                                "ignore_above": 1024
                            }
                        }
                    },
                    "alerting": {
                        "properties": {
                            "action_group_id": {
                                "type": "keyword",
                                "ignore_above": 1024
                            },
                            "action_subgroup": {
                                "type": "keyword",
                                "ignore_above": 1024
                            },
                            "instance_id": {
                                "type": "keyword",
                                "ignore_above": 1024
                            },
                            "outcome": {
                                "type": "keyword",
                                "ignore_above": 1024
                            },
                            "status": {
                                "type": "keyword",
                                "ignore_above": 1024
                            },
                            "summary": {
                                "properties": {
                                    "new": {
                                        "properties": {
                                            "count": {
                                                "type": "long"
                                            }
                                        }
                                    },
                                    "ongoing": {
                                        "properties": {
                                            "count": {
                                                "type": "long"
                                            }
                                        }
                                    },
                                    "recovered": {
                                        "properties": {
                                            "count": {
                                                "type": "long"
                                            }
                                        }
                                    }
                                }
                            }
                        }
                    },
                    "saved_objects": {
                        "type": "nested",
                        "properties": {
                            "id": {
                                "type": "keyword",
                                "ignore_above": 1024
                            },
                            "namespace": {
                                "type": "keyword",
                                "ignore_above": 1024
                            },
                            "rel": {
                                "type": "keyword",
                                "ignore_above": 1024
                            },
                            "space_agnostic": {
                                "type": "boolean"
                            },
                            "type": {
                                "type": "keyword",
                                "ignore_above": 1024
                            },
                            "type_id": {
                                "type": "keyword",
                                "ignore_above": 1024
                            }
                        }
                    },
                    "server_uuid": {
                        "type": "keyword",
                        "ignore_above": 1024
                    },
                    "space_ids": {
                        "type": "keyword",
                        "ignore_above": 1024,
                        "meta": {
                            "isArray": "true"
                        }
                    },
                    "task": {
                        "properties": {
                            "id": {
                                "type": "keyword"
                            },
                            "schedule_delay": {
                                "type": "long"
                            },
                            "scheduled": {
                                "type": "date"
                            }
                        }
                    },
                    "version": {
                        "type": "version"
                    }
                }
            },
            "log": {
                "properties": {
                    "level": {
                        "type": "keyword",
                        "ignore_above": 1024
                    },
                    "logger": {
                        "type": "keyword",
                        "ignore_above": 1024
                    }
                }
            },
            "message": {
                "type": "text",
                "norms": false
            },
            "rule": {
                "properties": {
                    "author": {
                        "type": "keyword",
                        "ignore_above": 1024,
                        "meta": {
                            "isArray": "true"
                        }
                    },
                    "category": {
                        "type": "keyword",
                        "ignore_above": 1024
                    },
                    "description": {
                        "type": "keyword",
                        "ignore_above": 1024
                    },
                    "id": {
                        "type": "keyword",
                        "ignore_above": 1024
                    },
                    "license": {
                        "type": "keyword",
                        "ignore_above": 1024
                    },
                    "name": {
                        "type": "keyword",
                        "ignore_above": 1024
                    },
                    "reference": {
                        "type": "keyword",
                        "ignore_above": 1024
                    },
                    "ruleset": {
                        "type": "keyword",
                        "ignore_above": 1024
                    },
                    "uuid": {
                        "type": "keyword",
                        "ignore_above": 1024
                    },
                    "version": {
                        "type": "keyword",
                        "ignore_above": 1024
                    }
                }
            },
            "tags": {
                "type": "keyword",
                "ignore_above": 1024,
                "meta": {
                    "isArray": "true"
                }
            },
            "user": {
                "properties": {
                    "id": {
                        "type": "keyword",
                        "ignore_above": 1024
                    },
                    "name": {
                        "type": "keyword",
                        "ignore_above": 1024,
                        "fields": {
                            "text": {
                                "type": "text",
                                "norms": false
                            }
                        }
                    }
                }
            }
        }
    }
}
```
</details>


<details>
<summary>This was the error message:</summary>

```
Detected an unhandled Promise rejection.

        Message: Cannot convert undefined or null to object

        Stack: TypeError: Cannot convert undefined or null to object
    at Function.keys (<anonymous>)
    at isMappingDefinition [...]
```
</details>

Reply:
I tested again with the new changes and I noticed something strange happening: when I reload the index list page and navigate to the Mappings details page of specific indices (for example, the one with the same mappings from my previous comment), the table view becomes empty. Eventually, it gets fixed after opening an index that doesn't cause this, but then when I reload the page again, I see an empty table view again. I don't see any error messages in the console. 🤔 




https://github.com/elastic/kibana/assets/59341489/4ad4bfc7-5be9-49bd-81e3-a60abf5034de

Reply:
Thank you @ElenaStoeva. I am able to generate the issue locally using the mapping that your provided. Let me try to fix this.

Reply:
@ElenaStoeva Looks like the [mapping](https://github.com/elastic/kibana/pull/178360#discussion_r1524744975) that you provided has some invalid keys. This is failing at this [line of verification](https://github.com/elastic/kibana/blob/main/x-pack/plugins/index_management/public/application/components/mappings_editor/lib/extract_mappings_definition.ts#L21). In this PR, we are using the treeview used template component.

For testing, please try to go component template and create a new component template using the json mapping you provided, you will see an error while creating one.

Reply:
Hi @saikatsarkar056, the mappings I provided are just the mappings of one of the hidden, auto-generated backing indices from data streams that are automatically created when you spin up Kibana, so I'm not sure why some of their mapping keys are not considered valid. 🤔 We should look into validation logic but, in the meantime, I think we could display the error message. However, wouldn't it be better if we use an [error prompt](https://eui.elastic.co/#/display/empty-prompt#loading-and-error-prompts) or an [error callout](https://eui.elastic.co/#/display/callout#danger) instead of simply a text?

Also, I see the error message displayed for some of these hidden indices, but I still see a blank page for others:

<img width="1705" alt="Screenshot 2024-03-18 at 13 57 43" src="https://github.com/elastic/kibana/assets/59341489/1a313b4f-af54-4a5f-9007-a53061bc1eca">

<details>
<summary>Mappings of this index:</summary>

```
{
    "mappings": {
        "_data_stream_timestamp": {
            "enabled": true
        },
        "dynamic_templates": [
            {
                "strings_as_keyword": {
                    "match_mapping_type": "string",
                    "mapping": {
                        "ignore_above": 1024,
                        "type": "keyword"
                    }
                }
            }
        ],
        "date_detection": false,
        "properties": {
            "@timestamp": {
                "type": "date"
            },
            "data_stream": {
                "properties": {
                    "dataset": {
                        "type": "constant_keyword",
                        "value": "deprecation.elasticsearch"
                    },
                    "namespace": {
                        "type": "constant_keyword",
                        "value": "default"
                    },
                    "type": {
                        "type": "constant_keyword",
                        "value": "logs"
                    }
                }
            },
            "ecs": {
                "properties": {
                    "version": {
                        "type": "keyword",
                        "ignore_above": 1024
                    }
                }
            },
            "elasticsearch": {
                "properties": {
                    "cluster": {
                        "properties": {
                            "name": {
                                "type": "keyword",
                                "ignore_above": 1024
                            },
                            "uuid": {
                                "type": "keyword",
                                "ignore_above": 1024
                            }
                        }
                    },
                    "elastic_product_origin": {
                        "type": "keyword",
                        "ignore_above": 1024
                    },
                    "event": {
                        "properties": {
                            "category": {
                                "type": "keyword",
                                "ignore_above": 1024
                            }
                        }
                    },
                    "http": {
                        "properties": {
                            "request": {
                                "properties": {
                                    "x_opaque_id": {
                                        "type": "keyword",
                                        "ignore_above": 1024
                                    }
                                }
                            }
                        }
                    },
                    "node": {
                        "properties": {
                            "id": {
                                "type": "keyword",
                                "ignore_above": 1024
                            },
                            "name": {
                                "type": "keyword",
                                "ignore_above": 1024
                            }
                        }
                    }
                }
            },
            "event": {
                "properties": {
                    "code": {
                        "type": "keyword",
                        "ignore_above": 1024
                    },
                    "dataset": {
                        "type": "keyword",
                        "ignore_above": 1024
                    }
                }
            },
            "host": {
                "properties": {
                    "ip": {
                        "type": "ip"
                    }
                }
            },
            "log": {
                "properties": {
                    "level": {
                        "type": "keyword",
                        "ignore_above": 1024
                    },
                    "logger": {
                        "type": "keyword",
                        "ignore_above": 1024
                    }
                }
            },
            "message": {
                "type": "text"
            },
            "process": {
                "properties": {
                    "thread": {
                        "properties": {
                            "name": {
                                "type": "keyword",
                                "ignore_above": 1024
                            }
                        }
                    }
                }
            },
            "service": {
                "properties": {
                    "name": {
                        "type": "keyword",
                        "ignore_above": 1024
                    }
                }
            },
            "trace": {
                "properties": {
                    "id": {
                        "type": "keyword",
                        "ignore_above": 1024
                    }
                }
            }
        }
    }
}
```
</details>

Reply:
@ElenaStoeva I used the EuiEmptyPrompt for displaying the error now.

<img width="908" alt="Screenshot 2024-03-18 at 9 43 30 AM" src="https://github.com/elastic/kibana/assets/132922331/e38b1169-1d20-410b-a1e3-4231b16bfa56">

Please use the latest commit. Now, you should not see black page for this [mapping](https://github.com/elastic/kibana/pull/178360#discussion_r1528693903).